crossopt/assignment4/countislands.py

```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python3
+
+
+# disjoint set union for grid components
+class _GridDSU:
+    def __init__(self, n, m):
+        self.parent = dict()
+        self.size = dict()
+        for i in range(n):
+            for j in range(m):
+                self.parent[(i, j)] = (i, j)
+                self.size[(i, j)] = 1
+
+    def __getitem__(self, u):
+        if self.parent[u] != u:
+            self.parent[u] = self[self.parent[u]]
+        return self.parent[u]
+
+    def join(self, u, v):
+        u = self[u]
+        v = self[v]
+        if self.size[u] < self.size[v]:
+            u, v = v, u
+        self.size[u] += self.size[v]
+        self.parent[v] = u
+
+
+class Grid(object):
+    def __init__(self, n, m, grid):
+        self.n = n
+        self.m = m
+        self.grid = grid
+
+    # returns DSU for grid where all islands are components
+    def colorIslands(self):
```

Comment:
Can you explain what you mean by "color" and why colors is somehow equivalent to DSU?
Do you think you can have a clearer name like MakeDsu?

Currently, there are 2 classes and 6 methods in total: and it is non-obvious what are the protocol or the contract between the two. Can you simplify this?

For example: if this method just returns the Dsu, then the feature might better placed on the __init__ of _GridDSU?
Currently Grid must know the _GridDSU constructor, the join method and the getter, it looks like if this method is moved to the _GridDSU constructor, then Grid only needs to know the _GridDSU constructor.